File: Data Structures & Algorithms/subsets/submission-7.go
func subsets(nums []int) [][]int {
    res := [][]int{}
    cur := []int{}

    dfs(0, len(nums), &res, cur, nums)
    return res
}

func dfs(i, j int, res *[][]int, cur, nums []int){

    if i == j{
        temp := make([]int, len(cur))
        copy(temp, cur)
        *res = append(*res, temp)
        return
    }

    cur = append(cur, nums[i])
    dfs(i+1, j, res, cur, nums)
    cur = cur[:len(cur)-1]
    dfs(i+1, j, res, cur, nums)
    return
}

File: Data Structures & Algorithms/combination-target-sum/submission-7.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        cur = []
        total = 0
        def dfs(i):
            nonlocal total
            if total == target:
                res.append(cur[:])
                return
            if i == len(nums) or total > target:
                return
            total += nums[i]
            cur.append(nums[i])
            dfs(i)
            total -= nums[i]
            cur.pop()
            dfs(i+1)
            return
        dfs(0)
        return res